lib/middleware.js

```diff
@@ -87,12 +87,25 @@ module.exports = function(self, options) {
 
       // If we're not down to one yet, winnow it down by prefix
       if ((candidates.length > 1) && self.prefixes) {
-        matches = req.url.match(/^\/[^/]+/);
-        if (matches) {
-          prefix = matches[0];
+        // If we're dealing with a locale which shares the hostname
+        // with other locales, but doesn't have a prefix, (i.e. it
+        // exists at the root level of the hostname), figure out
+        // that locale and set that as the current session's locale.
+        if (req.url === '/') {
+          // a locale which resides at the root level of a shared
+          // hostname won't have a prefix. So we select the candidate
+          // which doesn't have a prefix in this case.
           candidates = _.filter(candidates, function(candidate) {
-            return self.prefixes[candidate] === prefix;
+            return !_.has(self.prefixes, candidate);
           });
+        } else {
+          matches = req.url.match(/^\/[^/]+/);
+          if (matches) {
+            prefix = matches[0];
+            candidates = _.filter(candidates, function (candidate) {
+              return self.prefixes[candidate] === prefix;
+            });
```

Comment:
I don't see how this is going to do the right thing when there is no prefix (en or de in your example) and the URL isn't the root, but rather, let's say, /articles?

Reply:
Ah, thanks for the catch! This is an embarrassing miss from my end. Guess I got carried away by the home page edge cases since I was primarily having trouble with those pages. I pushed a fix just now!